rpcserver.go

```diff
@@ -2026,18 +2027,38 @@ func (r *rpcServer) parseOpenChannelReq(in *lnrpc.OpenChannelRequest,
 
 	case lnrpc.CommitmentType_ANCHORS:
 		channelType = new(lnwire.ChannelType)
-		*channelType = lnwire.ChannelType(*lnwire.NewRawFeatureVector(
+		fv := lnwire.NewRawFeatureVector(
 			lnwire.StaticRemoteKeyRequired,
 			lnwire.AnchorsZeroFeeHtlcTxRequired,
-		))
+		)
+
+		// The feature bit does not match the chan type.
+		if in.ZeroConf {
+			fv.Set(lnwire.ZeroConfChanType)
+		} else if in.ScidAlias {
+			// This is mutually exclusive with zero-conf.
+			fv.Set(lnwire.ScidAliasChanType)
+		}
```

Comment:
I'm having troubles understand this, it doesn't seem to match the spec. If you make them exclusive you cannot open 0-conf unannounced channels? The scid-alias is only mutually exclusive with announcing the channel, not with 0-conf.

I'd like to test this branch against eclair's implementation, but I couldn't find how to pass a `channel_type`  parameter to lnd's `openchannel` RPC, can you explain how that works? Even if I activate both features on the lnd node, it won't pick them up and open a 0-conf channel.

Reply:
Thanks, so I'll wait for this branch to be updated with your latest changes before I start doing cross-compatibility tests.

Reply:
I've updated the branch so you should be able to set both , lmk if you have any hiccups

Reply:
I've been able to successfully test the following cases where eclair is the funder:

* open a public 0-conf channel
* open a private channel with scid_alias without 0-conf
* open a private channel with scid_alias and 0-conf

All of these worked as expected, the only thing I couldn't test are payments from a remote node as lnd doesn't include routing hints by default.

I haven't been able to test the cases where lnd is the funder. I have activated `--protocol.zero-conf` and `--protocol.option-scid-alias` on the lnd node (and they show in `getinfo`), but even if I provided `--private`, `--zero_conf`, `--scid_alias` to the `openchannel` RPC lnd sends a `channel_type` that doesn't include any of these features.

Reply:
lnd has to be called with the `--private` flag with `AddInvoice` to include route hints
Are you using `lncli` binary with `open_channel --...` or the grpc `OpenChannel` call? I'm pretty sure that this failure is because lnd only sends the scid-alias or zero-conf channel_type with anchors unless it's like whitelisted (this carve-out is for LDK)

Reply:
I'm using `lncli` and eclair supports anchors (and advertises it in its `init`).

Reply:
So if I do `lncli openchannel <node> <amount> --channel_type=anchors --zero_conf` lnd will send a `channel_type` containing `anchors` + `zero_conf`?

Reply:
Yes it should, the integration tests I wrote cover this case except on the gRPC layer

Reply:
Great, thanks for the hint, that unblocked my tests! I can confirm that with this added argument, I'm able to open channels from lnd to eclair that use `scid_alias`, `zero_conf` or both and it's mostly working. I've only found two issues:

- when eclair doesn't send `zero_conf` in its `init`, but then tries to open a channel to lnd with `zero_conf` in the `channel_type`, lnd rejects it: I believe it would be useful to accept it, since the lnd node has `zero_conf` activated?
- when lnd opens a private channel with `scid_alias` (but without `zero_conf`) to eclair, lnd doesn't send a `channel_update` to eclair after 6 confirmations, so eclair is unable to create routing hints: that looks like a bug that should be fixed?

Reply:
The first bullet I don't think we'll support, it might be useful but kind of goes against our codebase with required features & having the feature bit

In the second, lnd should be sending a channel_update, but not after 6 confs. IIRC it is sent after min_confs with the on-chain SCID. I think eclair should be able to process the channel_update and in the route hints, use the alias + info in the channel_update. Here's what the spec says:
```
- MUST set the `short_channel_id` to either an `alias` it has
	  received from the peer, or the real channel `short_channel_id`.
```

Reply:
I think the issue could be that lnd (not specific to this PR) sends `channel_update` without waiting for the counterparty's `funding_locked`

Reply:
> I think the issue could be that lnd (not specific to this PR) sends channel_update without waiting for the counterparty's funding_locked

This would explain it. It really needs to be changed though, otherwise it simply cannot work. Think about your peer's state: if they haven't sent `channel_ready`, the real `short_channel_id` you're sending doesn't make sense to them, so the `channel_update` will be dropped. It would also really be much better to always use the alias since this is a private channel with `scid_alias`, which requires waiting for your peer's `channel_ready`...

Reply:
I'll look into fixing the funding_locked thing in a separate PR and rebasing this one on top. It's a long-standing issue and fixing it might break some tests

I can look into sending an alias instead, but it won't make it into this PR. It should be possible given the design I went with, but I don't want to make this PR any bigger

Reply:
I just tested, and lnd does indeed send its `channel_update` right after its `channel_ready`. But since it's received before we generate our own `channel_ready`, we drop that `channel_update`. This can be fixed in a separate PR.

Reply:
In the private channel case with eclair (which implements the spec properly, I *think* LDK has a carve-out), lnd would eventually send the `channel_update` after ~24 hours I think (terrible UX).

lnd-lnd channels are unaffected so if the LSP is an lnd-based node it's fine. I don't really know about other impls though. Tracking issue is here https://github.com/lightningnetwork/lnd/issues/3651

Reply:
> lnd-lnd channels are unaffected so if the LSP is an lnd-based node it's fine. I don't really know about other impls though. Tracking issue is here https://github.com/lightningnetwork/lnd/issues/3651

Can you detail how lnd-lnd can be unaffected? Does that means lnd accepts potentially junk `channel_update`s? Wouldn't that be a DoS vector? Or does lnd add an entry in its router after less than `min_conf` (but more than 1 conf) so that the `channel_update` isn't dropped?

Reply:
So lnd has an LRU cache that stores `channel_update`s it doesn't know of. Once lnd does know about the `channel_update` after seeing it on-chain after `min_confs`, it will check the cache for a matching `channel_update`, validate it and use it. If for some reason it's not in the cache, the channel recovers after ~24 hours due to a broadcast timer ticking. It's not ideal and leads to extra code, so I think sending `channel_update` only after `funding_locked` should let us remove it completely

Reply:
I've posted a draft fix for the above, since it will get merged on its own and then this PR will, I made a branch so you can test this branch with the patched channel_update logic: https://github.com/Crypt-iQ/lnd/tree/zero-conf-alias-patch-chanupdate

Reply:
Thanks, I just tested it and it does resolve the issue :+1: